README.md
# mdbook-chess

Create chess boards in your mdbooks!

![A chess board in starting position](./doc/demo_image.png)

# Installation

Simply:

```
cargo install mdbook-chess
```

And then add to your mdbook project as so:

```
# book.toml
[book]
authors = ["xd009642"]
language = "en"
multilingual = false
src = "src"
title = "mdbook-chess"

[preprocessor.chess]
```

For an example book just look to the provided [demo-book](./demo-book).

# Usage

Currently the chess boards are specified via a very simple YAML structure. With
5 possible fields:

* load - either "start" for the starting board or a FEN string. If not provided it's default starting board
* save - name to save board under (can be a list)
* moves - list of moves
* overwrite - boolean of whether to overwrite saved version of the board with these moves
* highlights - highlight a square on the board a list of squares
* lines - lines and arrows from square to square (currently only straight)

So we can create a starting board like using a code block with the language set to chess. Like:


    ```chess
    load: start
    ``` 

Or like:

    ```chess
    # mdbook seems to hate empty code blocks so have at least a blank line
    ```

_For brevity/formating the backticks and "chess" language specifier will be omitted
for future examples._

Of course if we want to refer to this board later we need to save it. So we can
create a board with a name to refer to it later as so:

```chess
save: our_board
```

We can then load it and apply some moves - this will overwrite it so to do
alternatives can set overwrite to false and start from previous position. We
can also save multiple boards from a point for exploring many different routes:

```chess
load: our_board
save: ["ck-main", "ck-2", "ck-3"]
moves: ["e4", "c6", "d4", "d5"]
```

Arrows are done by writing the two squares and the line between them, there are
currently 3 styles:

* `-` a line
* `->` a single direction arrow
* `<->` a bidirectional arrow

And we can use them as so:

```chess
load: our_board
moves: ["e4", "c6", "d4", "d5"]
lines: ["e2->e4", "d1<->h5", "a4-h4"]
```


Generally, if there's a syntax error in the markdown there's a reasonable
chance you'll see a bunch of SVG text dumped into the book. If this happens
read the log messages and hopefully an answer will be gleaned.

For an example of all the features in use please look to the demo book!

# Roadmap

## Potential future features

These are features I'd be willing to do if there was demand and they were
sufficiently easy or willing to accept PRs for.

* GIF generation
* L shaped move arrow for horsey
* Chess.com style icons (blunder, excellent etc)
* Styling and theming
* More layout options in page
* Rank and file labels
* Flip the board

## Unlikely features

These are features where they will only be added via PR but the PR may
not be accepted

* Chess variants
* Eval bar

# License and Copyright

All chess SVGs were created by Wikimedia user Cburnett and are CC BY-SA 3.0
licensed. You can find them [here](https://commons.wikimedia.org/wiki/Category:SVG_chess_pieces)
I've edited them to template in a transform and remove the outer tags for
ease of board generation.

All other aspects of this project are currently licensed under the terms of both
the MIT license and the Apache License (Version 2.0). See LICENSE-MIT and
LICENSE-APACHE for more details.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
 
The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.2.1] - 2023-07-18
### Fixed
- Fixed handling of non-text non-chess markdown elements (tables)

## [0.2.0] - 2022-09-04
### Added
- Square highlighting
- Implemented move arrows and expose via lines field in yaml

### Changed
- Only log chapters containing chess blocks
- Fixed squares on board so A1 is dark not light
 
## [0.1.0] - 2022-08-06
### Added
- Initial release. Specify boards and render to mdbook. 


src/lib.rs
use chess::{Board, ChessMove, Color, File, Piece, Rank, Square};
use mdbook::book::{Book, BookItem, Chapter};
use mdbook::errors::Error as MdBookError;
use mdbook::preprocess::{CmdPreprocessor, Preprocessor, PreprocessorContext};
use pulldown_cmark::{CodeBlockKind, CowStr, Event, Parser, Tag};
use pulldown_cmark_to_cmark::cmark;
use serde::Deserialize;
use std::collections::HashMap;
use std::path::Path;
use std::str::FromStr;
use std::{fmt, fs, io};

#[derive(Debug, Clone, Deserialize)]
pub struct BoardBlock {
    /// The name we want to refer to this board by as we mutate it
    name: Option<String>,
    /// This lets us go back to it in future
    checkpoint: Option<String>,
    /// If we're creating a new board
    board: Option<String>,
    /// Moves to apply to the board
    #[serde(default)]
    moves: Vec<String>,
}

const X_OFFSET: f32 = 0.6;
const Y_OFFSET: f32 = 0.3;
pub const PREPROCESSOR_NAME: &'static str = "mdbook-chess";

const fn get_board() -> &'static str {
    include_str!("../res/board.svg")
}

const fn get_piece(p: Piece, c: Color) -> &'static str {
    match (p, c) {
        (Piece::Pawn, Color::White) => include_str!("../res/P.svg"),
        (Piece::Knight, Color::White) => include_str!("../res/N.svg"),
        (Piece::Bishop, Color::White) => include_str!("../res/B.svg"),
        (Piece::Rook, Color::White) => include_str!("../res/R.svg"),
        (Piece::Queen, Color::White) => include_str!("../res/Q.svg"),
        (Piece::King, Color::White) => include_str!("../res/K.svg"),
        (Piece::Pawn, Color::Black) => include_str!("../res/p.svg"),
        (Piece::Knight, Color::Black) => include_str!("../res/n.svg"),
        (Piece::Bishop, Color::Black) => include_str!("../res/b.svg"),
        (Piece::Rook, Color::Black) => include_str!("../res/r.svg"),
        (Piece::Queen, Color::Black) => include_str!("../res/q.svg"),
        (Piece::King, Color::Black) => include_str!("../res/k.svg"),
    }
}

fn coordinate(file: File, rank: Rank) -> (f32, f32) {
    let rank = 70 - (rank.to_index() * 10);
    (
        (file.to_index() * 10) as f32 + X_OFFSET,
        rank as f32 + Y_OFFSET,
    )
}

pub fn generate_board(board: &Board) -> String {
    let mut pieces = String::new();
    for i in 0..64 {
        let square = unsafe { Square::new(i) };
        if let Some((piece, color)) = board.piece_on(square).zip(board.color_on(square)) {
            let (x, y) = coordinate(square.get_file(), square.get_rank());

            let piece_svg = get_piece(piece, color)
                .replace("$X_POSITION", &x.to_string())
                .replace("$Y_POSITION", &y.to_string());

            pieces.push_str(&piece_svg);
        }
    }

    get_board().replace("<!-- PIECES -->", &pieces)
}

pub fn save_board_from_fen(fen: &str, path: impl AsRef<Path>) -> io::Result<()> {
    let board = Board::from_str(fen)
        .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e.to_string()))?;
    save_board(&board, path)
}

/// Takes
pub fn save_board(board: &Board, path: impl AsRef<Path>) -> io::Result<()> {
    let svg = generate_board(&board);
    fs::write(path, svg)
}

pub fn run_preprocessor(ctx: &PreprocessorContext, mut book: Book) -> Result<Book, MdBookError> {
    // No settings so we'll skip
    book.for_each_mut(|item| {
        if let BookItem::Chapter(chapter) = item {
            let _ = process_code_blocks(chapter).map(|s| {
                chapter.content = s;
                eprintln!("chapter '{}' processed", chapter.name);
            });
        }
    });
    Ok(book)
}

fn process_code_blocks(chapter: &mut Chapter) -> Result<String, fmt::Error> {
    use CodeBlockKind::*;
    use CowStr::*;
    use Event::*;
    use Tag::{CodeBlock, Paragraph};

    let mut boards = HashMap::new();
    let mut checkpoints = HashMap::new();

    let mut output = String::with_capacity(chapter.content.len());
    let mut inside_block = false;
    let events = Parser::new(&chapter.content).map(|e| match (&e, inside_block) {
        (Start(CodeBlock(Fenced(Borrowed("chess")))), false) => {
            inside_block = true;
            Start(Paragraph)
        }
        (Text(Borrowed(text)), true) => {
            inside_block = false;
            Html(generate_board_svg(text, &mut boards, &mut checkpoints).into())
        }
        (End(CodeBlock(Fenced(Borrowed("chess")))), false) => End(Paragraph),
        _ => e,
    });

    cmark(events, &mut output).map(|_| output)
}

fn generate_board_svg(
    input: &str,
    boards: &mut HashMap<String, Board>,
    checkpoints: &mut HashMap<String, Board>,
) -> String {
    match serde_yaml::from_str::<BoardBlock>(input) {
        Ok(block) => {
            eprintln!("I'm doing this for ya: {:?}", block);
            let name = block.name.clone();
            let mut board = match block.board.as_deref() {
                Some("start") => Board::default(),
                Some(s) => match Board::from_str(s) {
                    Ok(b) => b,
                    Err(e) => {
                        eprintln!("Invalid FEN String: {}", e);
                        return get_board().to_string();
                    }
                },
                None => {
                    let res = name
                        .as_ref()
                        .and_then(|name| boards.get(name).or_else(|| checkpoints.get(name)))
                        .cloned();
                    match res {
                        Some(b) => b,
                        None => Board::default(),
                    }
                }
            };

            for m in &block.moves {
                match ChessMove::from_san(&board, m.as_str()) {
                    Ok(chess_move) => {
                        let new_board = board.make_move_new(chess_move);
                        board = new_board;
                    }
                    Err(_) => {
                        eprintln!("{} is an invalid SAN move", m);
                        return get_board().to_string();
                    }
                }
            }
            if let Some(s) = block.checkpoint.as_ref() {
                checkpoints.insert(s.clone(), board.clone());
            }
            if let Some(name) = name.clone() {
                boards.insert(name, board.clone());
            }
            generate_board(&board)
        }
        Err(e) => {
            eprintln!("Invalid YAML: {}", e);
            // We got nothing, lets just pop an empty board there:
            get_board().to_string()
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    pub fn ensure_svg_in_output() {
        let mut chapter = Chapter::new(
            "test",
            "```chess\nBoard: default\n```".to_string(),
            ".",
            vec![],
        );
        let s = process_code_blocks(&mut chapter).unwrap();
        assert!(s.contains(&generate_board(&Board::default())));
    }
}


src/main.rs
use chess_preproc::*;
use clap::{App, Arg, ArgMatches};
use mdbook::errors::Error;
use mdbook::preprocess::{CmdPreprocessor, Preprocessor};
use semver::{Version, VersionReq};
use std::{env, io, process};
use tracing::{info, trace, warn};
use tracing_subscriber::filter::EnvFilter;
use tracing_subscriber::{Layer, Registry};

pub mod arrows;
mod chess_preproc;

pub fn make_app() -> App<'static> {
    App::new(PREPROCESSOR_NAME)
        .about("A chess preprocessor which puts images of boards into your book")
        .subcommand(
            App::new("supports")
                .arg(Arg::new("renderer").required(true))
                .about("Check whether a renderer is supported by this preprocessor"),
        )
}

fn setup_logging() -> Result<(), Box<dyn std::error::Error>> {
    let filter = match env::var("RUST_LOG") {
        Ok(_) => EnvFilter::from_default_env(),
        _ => EnvFilter::new("mdbook_chess=info"),
    };

    let fmt = tracing_subscriber::fmt::Layer::default().with_writer(io::stderr);

    let subscriber = filter.and_then(fmt).with_subscriber(Registry::default());

    tracing::subscriber::set_global_default(subscriber)?;
    Ok(())
}

fn main() {
    if let Err(e) = setup_logging() {
        eprintln!("Failed to setup tracing logging: {}", e);
    }
    let matches = make_app().get_matches();

    if let Some(sub_args) = matches.subcommand_matches("supports") {
        handle_supports(sub_args);
    } else if let Err(e) = handle_preprocessing() {
        info!("{}", e);
        process::exit(1);
    }
}

fn handle_preprocessing() -> Result<(), Error> {
    let (ctx, book) = CmdPreprocessor::parse_input(io::stdin())?;

    let book_version = Version::parse(&ctx.mdbook_version)?;
    let version_req = VersionReq::parse(mdbook::MDBOOK_VERSION)?;

    if !version_req.matches(&book_version) {
        warn!(
            "Warning: The {} plugin was built against version {} of mdbook, \
             but we're being called from version {}",
            PREPROCESSOR_NAME,
            mdbook::MDBOOK_VERSION,
            ctx.mdbook_version
        );
    }
    info!("Processing!");

    let preproc = ChessPreprocessor;
    let processed_book = preproc.run(&ctx, book).expect("Failed to preprocess book");
    let s =
        serde_json::to_string(&processed_book).expect("Failed to convert processed book to json");
    println!("{}", s);
    Ok(())
}

fn handle_supports(sub_args: &ArgMatches) -> ! {
    let preproc = ChessPreprocessor;
    let renderer = sub_args.value_of("renderer").expect("Required argument");
    if preproc.supports_renderer(renderer) {
        process::exit(0);
    } else {
        process::exit(1);
    }
}


src/arrows.rs
use crate::*;
use chess::Square;
use serde::de::{self, Deserialize, Deserializer, Unexpected, Visitor};
use std::fmt;
use std::str::FromStr;

pub const ARROW_COLOUR: &'static str = "#008a00";
pub const ARROW_OPACITY: f32 = 0.50196078;

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct Line {
    pub start: Square,
    pub end: Square,
    pub head: Option<ArrowHead>,
}

#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
/// Type of arrow head
pub enum ArrowHead {
    /// An arrow like `start -> end`
    Single,
    /// An arrow like `start <-> end`
    Double,
}

impl<'de> Deserialize<'de> for Line {
    fn deserialize<D>(deserializer: D) -> Result<Line, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_str(ArrowVisitor)
    }
}

struct ArrowVisitor;

impl<'de> Visitor<'de> for ArrowVisitor {
    type Value = Line;

    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        write!(
            formatter,
            "a string containing two squares joined by a line or arrow"
        )
    }

    fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        let splits: Vec<&str> = s.split("-").collect();
        if splits.len() != 2 {
            Err(de::Error::invalid_value(Unexpected::Str(s), &self))
        } else {
            if splits[0].len() < 2 || splits[0].len() > 3 {
                Err(de::Error::invalid_value(Unexpected::Str(splits[0]), &self))
            } else if splits[1].len() < 2 || splits[1].len() > 3 {
                Err(de::Error::invalid_value(Unexpected::Str(splits[1]), &self))
            } else {
                let square_1 = &splits[0][..2];

                let square_1 = match Square::from_str(square_1) {
                    Ok(s) => s,
                    Err(_) => {
                        return Err(de::Error::invalid_value(Unexpected::Str(square_1), &self));
                    }
                };

                let square_2 = match splits[1].len() {
                    2 => splits[1],
                    3 => &splits[1][1..],
                    _ => unreachable!(),
                };

                let square_2 = match Square::from_str(square_2) {
                    Ok(s) => s,
                    Err(_) => {
                        return Err(de::Error::invalid_value(Unexpected::Str(square_2), &self));
                    }
                };

                let (invert, head) = match (splits[0].chars().nth(2), splits[1].chars().nth(0)) {
                    (Some('<'), Some('>')) => (false, Some(ArrowHead::Double)),
                    (Some('<'), _) => (true, Some(ArrowHead::Single)),
                    (None, Some('>')) => (false, Some(ArrowHead::Single)),
                    (Some(t), _) | (_, Some(t)) => {
                        if splits[0].len() != 2 && splits[1].len() != 2 {
                            return Err(de::Error::invalid_value(Unexpected::Char(t), &self));
                        } else {
                            (false, None)
                        }
                    }
                    (None, None) => (false, None),
                };

                if invert {
                    Ok(Line {
                        start: square_2,
                        end: square_1,
                        head,
                    })
                } else {
                    Ok(Line {
                        start: square_1,
                        end: square_2,
                        head,
                    })
                }
            }
        }
    }
}

impl Line {
    /// Generates an SVG string for the arrow. Currently relies on marker definitions written into
    /// board.
    pub fn svg_string(&self) -> String {
        if self.start == self.end {
            String::new()
        } else {
            let (mut x1, mut y1) = coordinate_from_square(&self.start);
            let (mut x2, mut y2) = coordinate_from_square(&self.end);
            const RETREAT: f32 = 4.0;

            // Move to centre of square;
            x1 += 5.0;
            y1 += 5.0;
            x2 += 5.0;
            y2 += 5.0;

            let norm = ((x2 - x1).powi(2) + (y2 - y1).powi(2)).sqrt();
            let t = RETREAT / norm;

            x1 = (1.0 - t) * x1 + t * x2;
            y1 = (1.0 - t) * y1 + t * y2;

            let t = (norm - RETREAT) / norm;
            x2 = (1.0 - t) * x1 + t * x2;
            y2 = (1.0 - t) * y1 + t * y2;

            let mut s = format!(
                r##"<line x1="{}" y1="{}" x2="{}" y2={} stroke="green" stroke-width="2.0" "##,
                x1, y1, x2, y2
            );

            match self.head.as_ref() {
                Some(ArrowHead::Single) => s.push_str(r#" marker-end="url(#endarrow)"/>"#),
                Some(ArrowHead::Double) => {
                    s.push_str(r#" marker-start="url(#startarrow)" marker-end="url(#endarrow)"/>"#)
                }
                None => s.push_str("/>"),
            }

            s
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde::Deserialize;

    #[derive(Deserialize)]
    struct TestStruct {
        lines: Vec<Line>,
    }

    #[test]
    fn arrow_deser() {
        let x = serde_json::from_str::<TestStruct>(
            r#"{"lines": ["d4->g6", "c3<-c4", "g7-h7", "a1<->a8"]}"#,
        )
        .unwrap();

        let expected = vec![
            Line {
                start: Square::D4,
                end: Square::G6,
                head: Some(ArrowHead::Single),
            },
            Line {
                start: Square::C4,
                end: Square::C3,
                head: Some(ArrowHead::Single),
            },
            Line {
                start: Square::G7,
                end: Square::H7,
                head: None,
            },
            Line {
                start: Square::A1,
                end: Square::A8,
                head: Some(ArrowHead::Double),
            },
        ];

        assert_eq!(x.lines, expected);
    }

    #[test]
    fn invalid_arrows() {
        assert!(serde_json::from_str::<TestStruct>(r#"{"lines": ["d4"]}"#).is_err());
        assert!(serde_json::from_str::<TestStruct>(r#"{"lines": ["d4-->g6"]}"#).is_err());
        assert!(serde_json::from_str::<TestStruct>(r#"{"lines": ["d4->>g6"]}"#).is_err());
        assert!(serde_json::from_str::<TestStruct>(r#"{"lines": ["d4<<->g6"]}"#).is_err());
        assert!(serde_json::from_str::<TestStruct>(r#"{"lines": ["d4->z6"]}"#).is_err());
        assert!(serde_json::from_str::<TestStruct>(r#"{"lines": ["d4->d20"]}"#).is_err());
        assert!(serde_json::from_str::<TestStruct>(r#"{"lines": ["d4z6"]}"#).is_err());
        assert!(serde_json::from_str::<TestStruct>(r#"{"lines": ["d4-7z6"]}"#).is_err());
        assert!(serde_json::from_str::<TestStruct>(r#"{"lines": ["d4=>z6"]}"#).is_err());
    }
}


src/chess_preproc.rs
use crate::arrows::*;
use chess::{Board, ChessMove, Color, File, Piece, Rank, Square};
use mdbook::book::{Book, BookItem, Chapter};
use mdbook::errors::Error as MdBookError;
use mdbook::preprocess::{Preprocessor, PreprocessorContext};
use pulldown_cmark::{CodeBlockKind, CowStr, Event, Parser, Tag};
use pulldown_cmark_to_cmark::cmark;
use serde::Deserialize;
use std::collections::HashMap;

use std::fmt;
use std::str::FromStr;
use tracing::{error, info};

/// A constant X axis offset to apply to all pieces (and pawns).
const X_OFFSET: f32 = 0.6;
/// A constant Y axis offset to apply to all pieces (and pawns).
const Y_OFFSET: f32 = 0.3;
/// The name of the plugin.
pub const PREPROCESSOR_NAME: &'static str = "mdbook-chess";

const fn true_value() -> bool {
    true
}

#[derive(Debug, Clone, Deserialize)]
pub struct BoardBlock {
    /// The name we want to refer to this board by as we mutate it
    load: Option<String>,
    /// This lets us go back to it in future
    save: Option<ManyOrOne>,
    /// If we're creating a new board
    board: Option<String>,
    /// Moves to apply to the board
    #[serde(default)]
    moves: Vec<String>,
    /// Whether to replace saved board with this update to it.
    #[serde(default = "true_value")]
    overwrite: bool,
    /// Squares to highlight
    #[serde(default)]
    highlights: Vec<String>,
    /// Arrows. Should be a list of string commands in the form ["a1->a2", "e2<->e4", "a3-g3"]
    #[serde(default)]
    lines: Vec<Line>,
}

pub struct ChessPreprocessor;

impl Preprocessor for ChessPreprocessor {
    fn name(&self) -> &str {
        PREPROCESSOR_NAME
    }

    fn run(&self, _ctx: &PreprocessorContext, mut book: Book) -> Result<Book, MdBookError> {
        // No settings so we'll skip
        book.for_each_mut(|item| {
            if let BookItem::Chapter(chapter) = item {
                let _ = process_code_blocks(chapter).map(|s| {
                    chapter.content = s;
                });
            }
        });
        Ok(book)
    }
}

/// To allow for `field = "hello"` or `field = ["hello", "world"]` in YAML
#[derive(Debug, Clone, Deserialize)]
#[serde(untagged)]
pub enum ManyOrOne {
    One(String),
    Many(Vec<String>),
}

impl ManyOrOne {
    fn iter<'a>(&'a self) -> Box<dyn Iterator<Item = &'a String> + 'a> {
        match self {
            Self::One(s) => Box::new(std::iter::once(s)),
            Self::Many(s) => Box::new(s.iter()),
        }
    }
}

impl BoardBlock {
    fn get_highlights(&self) -> Vec<Square> {
        self.highlights
            .iter()
            .filter_map(|x| {
                Square::from_str(x)
                    .map_err(|e| {
                        error!("Invalid square {}: {}", x, e);
                    })
                    .ok()
            })
            .collect()
    }
}

/// Returns the SVG for the board for us to embed.
const fn get_board() -> &'static str {
    include_str!("../res/board.svg")
}

/// Gets the templaed SVG path for pieces.
const fn get_piece(p: Piece, c: Color) -> &'static str {
    match (p, c) {
        (Piece::Pawn, Color::White) => include_str!("../res/P.svg"),
        (Piece::Knight, Color::White) => include_str!("../res/N.svg"),
        (Piece::Bishop, Color::White) => include_str!("../res/B.svg"),
        (Piece::Rook, Color::White) => include_str!("../res/R.svg"),
        (Piece::Queen, Color::White) => include_str!("../res/Q.svg"),
        (Piece::King, Color::White) => include_str!("../res/K.svg"),
        (Piece::Pawn, Color::Black) => include_str!("../res/p.svg"),
        (Piece::Knight, Color::Black) => include_str!("../res/n.svg"),
        (Piece::Bishop, Color::Black) => include_str!("../res/b.svg"),
        (Piece::Rook, Color::Black) => include_str!("../res/r.svg"),
        (Piece::Queen, Color::Black) => include_str!("../res/q.svg"),
        (Piece::King, Color::Black) => include_str!("../res/k.svg"),
    }
}

const fn square_highlight() -> &'static str {
    include_str!("../res/highlight.svg")
}

pub fn coordinate_from_square(square: &Square) -> (f32, f32) {
    coordinate(square.get_file(), square.get_rank())
}

#[inline(always)]
/// For a given rank and file return (x, y) coordinate.
pub fn coordinate(file: File, rank: Rank) -> (f32, f32) {
    let rank = 70 - (rank.to_index() * 10);
    (
        (file.to_index() * 10) as f32 + X_OFFSET,
        rank as f32 + Y_OFFSET,
    )
}

/// Given a board layout generates an SVG string for the board
pub fn generate_board(board: &Board, highlights: Option<Vec<Square>>, lines: &[Line]) -> String {
    let mut pieces = String::new();
    for i in 0..64 {
        let square = unsafe { Square::new(i) };
        if let Some((piece, color)) = board.piece_on(square).zip(board.color_on(square)) {
            let (x, y) = coordinate_from_square(&square);

            let piece_svg = get_piece(piece, color)
                .replace("$X_POSITION", &x.to_string())
                .replace("$Y_POSITION", &y.to_string());

            pieces.push_str(&piece_svg);
        }
    }

    if let Some(highlights) = highlights {
        for square in highlights.iter() {
            let (x, y) = coordinate_from_square(square);
            let square = square_highlight()
                .replace("$X_POSITION", &(x - 0.6).to_string())
                .replace("$Y_POSITION", &(y - 0.27).to_string());
            pieces.push_str(&square);
        }
    }

    for line in lines {
        pieces.push_str(&line.svg_string());
    }

    get_board().replace("<!-- PIECES -->", &pieces)
}

/// Generate new markdown for a chapter
fn process_code_blocks(chapter: &mut Chapter) -> Result<String, fmt::Error> {
    use CodeBlockKind::*;
    use CowStr::*;
    use Event::*;
    use Tag::{CodeBlock, Paragraph};

    let mut boards = HashMap::new();

    let mut logged_found = false;
    let mut output = String::with_capacity(chapter.content.len());
    let mut inside_block = false;
    let events = Parser::new(&chapter.content).map(|e| match (&e, inside_block) {
        (Start(CodeBlock(Fenced(Borrowed("chess")))), false) => {
            inside_block = true;
            if !logged_found {
                info!("Found chess block(s) in {}", chapter.name);
                logged_found = true;
            }
            Start(Paragraph)
        }
        (Text(Borrowed(text)), true) => {
            inside_block = false;
            Html(process_chess_block(text, &mut boards).into())
        }
        (End(CodeBlock(Fenced(Borrowed("chess")))), false) => End(Paragraph),
        (Text(text), false) => Html(text.clone()),
        _ => e,
    });

    cmark(events, &mut output).map(|_| output)
}

/// Given our c
fn process_chess_block(input: &str, boards: &mut HashMap<String, Board>) -> String {
    match serde_yaml::from_str::<BoardBlock>(input) {
        Ok(block) => {
            let name = block.load.clone();
            let mut board = match block.board.as_deref() {
                Some("start") => Board::default(),
                Some(s) => match Board::from_str(s) {
                    Ok(b) => b,
                    Err(e) => {
                        error!("Invalid FEN String: {}", e);
                        return get_board().to_string();
                    }
                },
                None => {
                    let res = name.as_ref().and_then(|name| boards.get(name)).cloned();
                    match res {
                        Some(b) => b,
                        None => Board::default(),
                    }
                }
            };

            for m in &block.moves {
                match ChessMove::from_san(&board, m.as_str()) {
                    Ok(chess_move) => {
                        let new_board = board.make_move_new(chess_move);
                        board = new_board;
                    }
                    Err(_) => {
                        error!("{} is an invalid SAN move", m);
                        return get_board().to_string();
                    }
                }
            }
            if let Some(s) = block.save.as_ref() {
                for save in s.iter() {
                    boards.insert(save.clone(), board.clone());
                }
            }
            if block.overwrite && !matches!(name.as_deref(), Some("start")) {
                if let Some(name) = name.clone() {
                    boards.insert(name, board.clone());
                }
            }
            generate_board(&board, Some(block.get_highlights()), &block.lines)
        }
        Err(e) => {
            error!("Creating default board invalid YAML: {}", e);
            // We got nothing, lets just pop a default board
            generate_board(&Board::default(), None, &[])
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ensure_svg_in_output() {
        let mut chapter = Chapter::new(
            "test",
            "```chess\nBoard: default\n```".to_string(),
            ".",
            vec![],
        );
        let s = process_code_blocks(&mut chapter).unwrap();
        assert!(s.contains(&generate_board(&Board::default(), None, &[])));
    }

    #[test]
    fn dont_remove_tables() {
        let mut chapter = Chapter::new(
            "test",
            "|foo|bar|\n|---|---|\n|a|b|".to_string(),
            ".",
            vec![],
        );
        let s = process_code_blocks(&mut chapter).unwrap();
        assert!(!s.contains(r#"\|foo|bar|"#), "{}", s);
    }
}
